### DISCUSSION TITLE
Can ArgoCd deploy application without using the git repo as source?

### DISCUSSION THREAD

**Discussion by ssharma9125** [argoproj/argo-cd#18867] [Q&A]
In my organization, I have a ci/cd process for deploying the applications with jenkins. In my ci/cd pipeline the helm chart for a specific service is generated dynamically using the environment values and secrets stored in the postgres database, values are fetched from this database and then the chart is generate and applied to the targeted cluster using the helm install or helm upgrade command and after the services are deployed the generated kubernetes manifests are not stored in any helm or git repo. 

Now I want to use the arog-cd for the CD part of the pipeline, so is it possible to deploy the application while still maintaining the current architecture of the infrastructure (not storing the k8s manifests in repo) and just applying the dynamically generated manifests to the cluster through argo-cd?

**Comment by ChristianCiach** [comment]
That sounds doable, but with some caveats.

I think there is no other option than to configure your argo-cd application with a valid git repository as a source, *but* this git repository could be almost empty. You could probably even use the same (empty) git repository for all of your applications.

To generate your kubernetes manifests you could then write your own [content management plugin ](https://argo-cd.readthedocs.io/en/stable/operator-manual/config-management-plugins/). Using such a plugin you could do virtually anything you like, as long as the output is a set of valid kubernetes manifests. 

Caveats:

* Argo-CD checks the git repository for changes every 3 minutes (by default). But in your case your git repo would almost never change, so your Content Management Plugin would only be triggered every 24h (hard refresh). If you cannot commit changes to your git repo every time your manifests change, you probably want to configure `--repo-cache-expiration` with a lower value to force the re-generation of your manifests at a lower interval. 
* You say you are using Helm for deployment, so please keep in mind that Argo-CD doesn't use Helm for package management. Argo-CD only ever uses `helm template` to render the YAML and then uses kubectl for actual deployment. Support for Helm-hooks in Argo-CD is incomplete, so please keep that in mind.

**Reply by ChristianCiach** [reply]
But I would probably do this another way. I would probably run a mini git-server inside the target cluster. Instead of letting your existing pipeline directly deploy to your kubernetes cluster, your pipeline should push the generated manifests into this "sidecar" git-repo of argo-cd. Argo-CD would then monitor this repo, without any plugins.

See this comment by a lead maintainer of Argo-CD:

* https://github.com/argoproj/argo-cd/discussions/9912#discussioncomment-3107833

**Reply by ssharma9125** [reply]
Thanks for the reply! for the mini git server approach I have multiple clusters running for multiple projects so in this approach, I'll need to run the git server on every cluster?

**Reply by ChristianCiach** [reply]
Well, I didn't really think this through. I would probably try to run the git server in the same cluster as your Argo-CD instance, so that argo-cd can use the local url of a kubernetes Service to connect to the repo.

**Reply by ssharma9125** [reply]
correct me if I am wrong, you are saying that i should push the manifests of all the services of all the cluster in the single git repo server with argo running on the same cluster?

**Reply by ChristianCiach** [reply]
Well, I am not trying to present you with a complete solution :) Just pointing out options.

You could also run a separate argo-cd instance for every target cluster, but this would probably make using the UI a lot harder and comes with a lot of maintenance overhead. See [How many do you need? Argo CD Architectures Explained](https://akuity.io/blog/argo-cd-architectures-explained/)

I don't know what kind of architecture fits the best for your situation. Argo-CD is unopinionated about this on purpose.

**Comment by ChristianCiach** [comment]
Argo-CD can also use Helm-Charts (https and OCI artifacts) as sources.

There is even a proposal to support OCI artifacts containing any kind of Kubernetes manifests some time in the future:

* https://github.com/argoproj/argo-cd/issues/17564

There is no time frame for this (but there is a WIP pull request, so I am hopeful that this is actually coming eventually). 

Maybe this is of some use for you?

**Reply by ssharma9125** [reply]
I'll give it a try. Thank you for your help

**Comment by SleepyBrett** [comment]
I'm looking for a crd i can place on the cluster that says 'such and such a chart, such and such a target and such and such values'. Essentially approximately the way that the crossplane helm-provider uses, however the crossplane helm provider doesn't have great feedback mechanisms. It can tell that the helm chart was applied but won't tell me if the pods are failing to start, etc.
